### DISCUSSION TITLE
Jump to next instance of variable

### DISCUSSION THREAD

**Discussion by evandavis** [joaotavora/eglot#1361] [Ideas]
Would it be possible to add a method I could bind to a key to jump between occurrences of a variable?  In [this discussion on the LSP package](https://github.com/emacs-lsp/lsp-mode/issues/821) there is a suggestion to hook into the `textDocument/documentHighlight` set to allow navigation between occurrences; could something like that work in eglot?

I understand I can mostly do this using `isearch-forward-symbol-at-point`, but that will match other stuff if the token happens to be a substring of another token.

I'd also be happy implementing this outside the package if someone could show me how :) but a quick keybinding to go from an import line to "where exactly is this used?" seems like something more people might find useful. Thanks for your consideration!

**Comment by joaotavora** [comment]
I don't have time to give you more advice than: look in the source code.  Seems like a good idea anyway :+1: , and `textDocument/documentHightlight` seems reasonable.

**Comment by nemethf** [comment]
I think `M-?` collects the requested occurrences for some LSP backends, then `next-error` and `previous-error` can be used to navigate among the results.  If this is not enough, you can customize `xref-show-definitions-function` and friends to your own function.

However, sometimes navigating among the instances of a variable is not useful enough.  See this feature request: https://github.com/rust-lang/rust-analyzer/issues/16365

**Reply by evandavis** [reply]
I use `xref-find-references` (`M-?`) all the time and love it; I'm specifically looking for a single chord to jump to the next reference within the current buffer.

**Reply by nemethf** [reply]
But `next-error` and `previous-error` are already bound to `M-g n` and `M-g p`.

**Reply by joaotavora** [reply]
I guess @evandavis wants something quicker that doesn't require an explicit `xref-find-references` call and piggy backs on the existing highlights of `textDocument/documentHighlight`

**Reply by evandavis** [reply]
@nemethf `xref-find-references` opens an xref buffer with references across the project and also loses which specific occurrence was at point at the time of call. So, like `isearch-forward-symbol-at-point`, it _kinda_ works, but takes at a minimum two chords and in most cases takes manual intervention to make sure I'm landing on the right thing. 

What I'm looking for is a single chord for each of previous/next, maybe `M-g j` and `M-g k`, to take me immediately to the previous/next occurrence of the symbol at point.

**Reply by nemethf** [reply]
OK.  I still think that everything is there in the xref realm.  So someone can create the requested feature by implementing a xref-show-xrefs-function/xref-show-definitions-function.

**Reply by joaotavora** [reply]
@nemethf you're probably right. In fact, this feature (quick jump not requiring `xref-find-references) could be implemented more powerfully like that, perhaps even more robustly: no dependence on the highlight capability or funky synchronization wait-for-highlight issues.  A tad less efficient, since the same info is being requested twice in some cases.

I say "probably" because it could be that `xref-find-references` which relies on `textDocument/references` doesn't return the same stuff that `textDocument/documentHighlight` returns.  But that could happen either way.
